code/game/machinery/computer/camera.dm

```diff
@@ -4,89 +4,140 @@
+					break
+		if(!(user in watchers))
+			user.unset_machine() // no usable camera on the network, we disconnect the user from the computer.
 			return
+	use_camera_console(user)
 
-		var/list/L = list()
-		for (var/obj/machinery/camera/C in cameranet.cameras)
-			if((z > ZLEVEL_SPACEMAX || C.z > ZLEVEL_SPACEMAX) && (C.z != z))//if on away mission, can only recieve feed from same z_level cameras
-				continue
-			L.Add(C)
-
-		camera_sort(L)
-
-		var/list/D = list()
-		D["Cancel"] = "Cancel"
-		for(var/obj/machinery/camera/C in L)
-			if(!C.network)
-				spawn(0)
-					throw EXCEPTION("Camera in a cameranet has no camera network")
-				continue
-			if(!(istype(C.network,/list)))
-				spawn(0)
-					throw EXCEPTION("Camera in a cameranet has a non-list camera network")
-				continue
-			var/list/tempnetwork = C.network&network
-			if(tempnetwork.len)
-				D[text("[][]", C.c_tag, (C.status ? null : " (Deactivated)"))] = C
-
-		var/t = input(user, "Which camera should you change to?") as null|anything in D
-		if(!t)
-			user.unset_machine()
-			return 0
+/obj/machinery/computer/security/proc/use_camera_console(mob/user)
+	var/list/camera_list = get_available_cameras()
+	var/t = input(user, "Which camera should you change to?") as null|anything in camera_list
+	if(user.machine != src) //while we were choosing we got disconnected from our computer or are using another machine.
+		return
+	if(!t)
+		user.unset_machine()
+		return
 
-		var/obj/machinery/camera/C = D[t]
+	var/obj/machinery/camera/C = camera_list[t]
 
-		if(t == "Cancel")
+	if(t == "Cancel")
+		user.unset_machine()
+		return
+	if(C)
+		var/camera_fail = 0
+		if(!C.can_use() || user.machine != src || user.eye_blind || user.incapacitated())
+			camera_fail = 1
+		else if(isrobot(user))
+			var/list/viewing = viewers(src)
+			if(!viewing.Find(user))
+				camera_fail = 1
+		else if(!issilicon(user))
+			if(!Adjacent(user))
+				camera_fail = 1
+
+		if(camera_fail)
 			user.unset_machine()
 			return 0
 
-		if(C)
-			if ((get_dist(user, src) > 1 || user.machine != src || user.eye_blind || !( C.can_use() )) && (!istype(user, /mob/living/silicon/ai)))
-				user.unset_machine()
-				if(!C.can_use() && !isAI(user))
-					src.current = null
-
-				return 0
-			else
-				if(isAI(user))
-					var/mob/living/silicon/ai/A = user
-					A.eyeobj.setLoc(get_turf(C))
-					A.client.eye = A.eyeobj
-				else
-					src.current = C
-					use_power(50)
-
-				spawn(5)
-					attack_hand(user)
-		return
-
-
+		if(isAI(user))
+			var/mob/living/silicon/ai/A = user
+			A.eyeobj.setLoc(get_turf(C))
+			A.client.eye = A.eyeobj
+		else
+			user.reset_perspective(C)
+		watchers[user] = C
+		use_power(50)
+		spawn(5)
+			use_camera_console(user)
+	else
+		user.unset_machine()
+
+//returns the list of cameras accessible from this computer
+/obj/machinery/computer/security/proc/get_available_cameras()

```

Comment:
I created this proc and proc/use_camera_console() to reduce the size of computer/security/attack_hand()